[MLIR][Linalg] Fix insert_slice fusion with rank reduction

RoboTux opened the pull request:
Insert_slice fusion with a linalg producer does not account for
possible rank-reduction in the insert_slice return type. When that
happens, a tensor.cast gets generated due to the type mismatch which is
invalid for tensor with different rank. This later trips other pass.

RoboTux committed:
[MLIR][Linalg] Fix insert_slice fusion with rank reduction

Insert_slice fusion with a linalg producer does not account for
possible rank-reduction in the insert_slice return type. When that
happens, a tosa.cast gets generated due to the type mismatch which is
invalid for tensor with different rank. This later trips other pass.

RoboTux commented:
Ping?

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Fusion.cpp:
```diff
@@ -235,6 +237,31 @@ mlir::linalg::fuseProducerOfTensor(OpBuilder &b, OpOperand &consumerOpOperand) {
+                                   const llvm::SmallBitVector &dropDims) {
+  auto fromType = cast<ShapedType>(from.getType());
+  assert(fromType.getRank() == dropDims.size());
+  SmallVector<ReassociationIndices, 2> reassocIdxsVec;
+  ReassociationIndices reassocIdxs;
```
Could you document these two?

banach-space reviewed (commented):
Looks good overall, I've made some minor suggestions. I am surprised we didn't hit this before.

Thanks for the contribution Thomas and apologies for the delay reviewing 🙏🏻

RoboTux committed:
Add more comments and simplify test

RoboTux committed:
Fix clang-format

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Fusion.cpp:
```diff
@@ -272,12 +308,16 @@ mlir::linalg::fuseProducerOfTensor(OpBuilder &b, OpResult producerOpResult,
   // Replace use.
+  Value def = fusedProducer->getResult(producerOpResult.getResultNumber());
+  Type consumerType = consumerOpOperand.get().getType();
+  // Rank-reduction occured as part of the extract_slice.
```
I have just realised ... Why not re-use rank-reducing `tensor.extract_slice` instead? It's a bit counter-intuitive to make that a rank-reducing `tensor.extract_slice` is replaced with a pair of `tensor.extract_slice` + `tensor.collapse_shape`.

RoboTux replied:
The extract_slice is done on the producer's arguments while the expand_shape is done on its result. I did look into doing the rank-reducing on the argument since the result type is derived from the DPS init, but the pass applies to both named and generic linalg ops which makes this tricky. In the case of a generic, one of the input might be using the unit dimension in a non-dim affine expression which requires changing the affine maps to update, but since the pass works on the LinalgOp interface there is no way to change the affine map (since named ops have an implicit map). Of course one could do an if but the code becomes quite complex.

My feeling this is the work for another pass. With this one, the producing linalg is moved inside the loop, alongside the consuming generic. Then it's a matter of folding expand shape / extract slice which sounds like a simple linalg fusion that probably already exists.

RoboTux replied:
Happy for me to close this?

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Fusion.cpp:
```diff
@@ -235,6 +237,39 @@ mlir::linalg::fuseProducerOfTensor(OpBuilder &b, OpOperand &consumerOpOperand) {
+  assert(fromType.getRank() == dropDims.size() &&
+         "dropDims dimension does not match from tensor rank");
+  // Computed reassociation map for the corresponding tensor.collapse_shape.
+  SmallVector<ReassociationIndices, 2> reassocIdxsVec;
```
OK, thanks for the comments! Initially I got a bit confused -  in my "head" these are reassociation groups. As in, this is _a vector of reassociation groups_. I would rename this to make it clearer (perhaps there's a better name?)

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Fusion.cpp:
```diff
@@ -235,6 +237,39 @@ mlir::linalg::fuseProducerOfTensor(OpBuilder &b, OpOperand &consumerOpOperand) {
   return fuseProducerOfTensor(b, producerOpResult, consumerOpOperand);
 }
 
+/// Create tensor.collapse_shape to drop dimensions in `dropDims` in tensor
+/// `from`.
```
Since this is only used for dropping unit dims ("rank reducing extract slice"), I would make that clear and also add an assert. Generalise is fine, but then we should test more.

banach-space reviewed (commented):
Thanks for the updates, Thomas!

I've left a few smallish comments, but before addressing those, it would be good to make sure that we do indeed need `tensor.collapse_shape` (as opposed to _a rank reducing_ `tensor.extract_slice`). I am sure that you thought it through and there's a good reason to insert `collapse_shape`, but it's not obvious to me just yet 😅

RoboTux commented:
Thanks for the review @banach-space , I hope the new revision addresses all your comment in a satisfactory way.

RoboTux commented:
> Thanks for the updates, Thomas!
> 
> I've left a few smallish comments, but before addressing those, it would be good to make sure that we do indeed need `tensor.collapse_shape` (as opposed to _a rank reducing_ `tensor.extract_slice`). I am sure that you thought it through and there's a good reason to insert `collapse_shape`, but it's not obvious to me just yet 😅

That's a very good question and I'll need to get back to you before I can answer fully. To be clear, the fusion does not insert a new extract_slice, this is done by another transform. I don't even think it was there in the earlier example and I don't know why there's a new linalg as well so I will look into understanding that. Normally there is an extract slice that might happen to collapse some unit dimension which then causes a type mismatch. This patch is to ensure we do the proper collapse_shape after the extract_slice is fused to reproduce that unit dimension collapsing and make the type match. By the way the if is because the pass requires the producing linalg and consuming linalg (the one that consume the extract_slice) to be in different block, the bug I was facing was a case of loop but I used a if here to make the code simpler.

RoboTux commented:
My bad, the pass does indeed create an extract slice (or a subview if working on memref). The code that does that is in Linalg Utils and doesn't deal with dropped unit dimension. I'll rework the patch along those lines instead.

RoboTux commented:
Sorry for the long pause. I've restarted working on it and should have a patch soon.

RoboTux committed:
Merge remote-tracking branch 'upstream/main' into fix_rank-reduced_insert_slice_fusion

RoboTux committed:
Address comments

- rename collapseTo to better reflect its usage
- assert it only collapse unit dimensions
- rename ReassociationIndices-using variables to reassocGroup and
  reassocMaps, the same terminology used in tensor.collapse_shape
  documentation
- use more representative test with comments to better explain what the
  patch does

RoboTux commented:
> > Thanks for the updates, Thomas!
> > I've left a few smallish comments, but before addressing those, it would be good to make sure that we do indeed need `tensor.collapse_shape` (as opposed to _a rank reducing_ `tensor.extract_slice`). I am sure that you thought it through and there's a good reason to insert `collapse_shape`, but it's not obvious to me just yet 😅
> 
> That's a very good question and I'll need to get back to you before I can answer fully. To be clear, the fusion does not insert a new extract_slice, this is done by another transform. I don't even think it was there in the earlier example and I don't know why there's a new linalg as well so I will look into understanding that. Normally there is an extract slice that might happen to collapse some unit dimension which then causes a type mismatch. This patch is to ensure we do the proper collapse_shape after the extract_slice is fused to reproduce that unit dimension collapsing and make the type match. By the way the if is because the pass requires the producing linalg and consuming linalg (the one that consume the extract_slice) to be in different block, the bug I was facing was a case of loop but I used a if here to make the code simpler.

This is answered in my other comment: https://github.com/llvm/llvm-project/pull/130961#discussion_r2077407969 but I'll rephrase it here in case that comment wasn't clear enough.

First of all, the fusion is worth doing in case of rank-reducing extract_slice even if an expand_shape is involved because the producer is moved inside the loop and operate directly on rank-reduced buffers. The only case it wouldn't be worth it is if the extract_slice was only used to do rank-reduction but that would be strange IR in my opinion, a collapse_shape should have been used instead in that case.

Second the collapse_shape is used because it would add a lot of complexity to fold the rank-reduction in the extract_slice operations of the producer's operands. This is because the producing linalg could be something like:

```
#map0 = affine_map<(d0, d1) -> (d0, d1)
#map1 = affine_map<(d0, d1) -> (d0, d0 + d1)
linalg.generic
    {indexing_maps = [#map0, #map1], iterator_types = ["parallel", "parallel"]}
    ins(%0, %1 : tensor<1x4xf32>, tensor <1x4xf32>) outs(%2 : tensor<1x4>) {
^bb0(%in : f32, %in1 : f32):
    %result = arith.addf %in, %in1 : f32
    linalg.yield %result : f32
}
// followed by loop with consumer
```

If the dimension #1 was used in a loop of step 2 and dimension #0 was rank-reduced, one would at the very least need to removed the dropped dimension from the indexing_maps as follows:

```
#map0 = affine_map<(d1) -> (d1)
#map1 = affine_map<(d1) -> (0 + d1)

// following is inside the loop where the consumer is
%2 = tensor.empty() : tensor<2xf32>
%3 = tensor.extract_slice %0 into %2[1,%iv][1,2][1,1] : tensor<2xf32>
%4 = tensor.empty() : tensor<2xf32>
%5 = tensor.extract_slice %1 into %4[1,%iv][1,2][1,1] : tensor<2xf32>
linalg.generic
    {indexing_maps = [#map0, #map1], iterator_types = ["parallel", "parallel"]}
    ins(%3, %5 : tensor<2xf32>, tensor <2xf32>) outs(%2 : tensor<2xf32>) {
^bb0(%in : f32, %in1 : f32):
    %result = arith.addf %in, %in1 : f32
    linalg.yield %result : f32
}
```

However the pass does not operate on GenericOp but on LinalgOp which does not allow modifying indexing maps since it can be an interface for a named linalg op with an implicit indexing maps. So one would need to distinguish between GenericOp and other LinalgOp. I also feel that this is a further optimization to fold the rank-reducing earlier, better suited into a dedicated pass if there isn't already one.

Does that make sense to you?

banach-space commented on mlir/test/Dialect/Linalg/tile-and-fuse-tensors.mlir:
```diff
@@ -318,3 +318,81 @@ func.func @pad_generic_static(%small_input: tensor<58x1xf32>, %large_input: tens
+  %c0 = arith.constant 0 : index
+  %c2 = arith.constant 2 : index
+  %c6 = arith.constant 6 : index
+  %cst = arith.constant 0.0 : f32
+  %init1 = tensor.empty() : tensor<1x6x6xf32>
+  %fill1 = linalg.fill ins(%cst : f32) outs(%init1 : tensor<1x6x6xf32>) -> tensor<1x6x6xf32>
```
This could be replaced with function input parameters.

banach-space reviewed (commented):
To me this makes sense, thanks for the discussion and apologies for the delay!

I've left a few comments inline, but I also have two higher-level suggestion:
* Move `dropGivenUnitDims` to mlir/lib/Dialect/Tensor/Utils.cpp -> this hook is a fairly generic helper.
* The test that's added here is quite complex - could you try simplifying it? In principle, the only thing that matters is "rank reduction", everything else (e.g. other ops) is creating noise. I've made some suggestions inline.

Let me also ping @qedawkins and @javedabsar , who are experts in this area :)

RoboTux committed:
Clean up code

dropGivenUnitDims():
- move assert out of loop
- rework algorithm to make grouping more explicit and avoid complex
  nested ifs
- fix occured typo

Test: remove all tensor.empty and linalg.fill

RoboTux committed:
Fix codestyle

RoboTux commented:
Sorry I forgot to move the function into Utils.cpp, will do.

RoboTux committed:
Move dropGivenUnitDims to Tensor Utils

banach-space reviewed (commented):
Thanks Thomas, final suggestions from me.

RoboTux commented on mlir/lib/Dialect/Tensor/Utils/Utils.cpp:
```diff
@@ -94,6 +94,39 @@ mlir::tensor::computeTransposedType(RankedTensorType rankedTensorType,
   return transposedTensorType;
 }
 
+/// Create tensor.collapse_shape to drop unit dimensions in `dropDims` in tensor
+/// `from`.
```
Doh, I meant to remove that after moving the code there but forgot.

RoboTux committed:
Address review comments

Utils:
- drop comments on implementation
- rename from into src

Fusion:
- restrict live range of droppedDims
- clarify comment for rank-reduction check

Test:
- Use more descriptive SSA and FileCheck variables
- Emphasize the rank-reducing extract_slice in the input IR as the key
  aspect of the test.

banach-space reviewed (approved):
Great, thanks for all the updates, LGTM!

With no comments from other reviewers, I think that this is safe to land. You may want to wait till tomorrow - just in case somebody changes their mind  and decides to review :) (given that this has been in review for a while, I suspect that you are not in a rush)

I really enjoyed the discussion, Thomas! Send more PRs :)
